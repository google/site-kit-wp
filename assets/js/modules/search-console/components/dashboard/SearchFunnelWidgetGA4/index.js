/**
 * SearchFunnelWidgetGA4 component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { isURL } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import {
	MODULES_SEARCH_CONSOLE,
	DATE_RANGE_OFFSET,
} from '../../../datastore/constants';
import { CORE_SITE } from '../../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import PreviewBlock from '../../../../../components/PreviewBlock';
import Header from './Header';
import Footer from './Footer';
import Overview from './Overview';
import { CORE_MODULES } from '../../../../../googlesitekit/modules/datastore/constants';
import useViewOnly from '../../../../../hooks/useViewOnly';
import {
	MODULES_ANALYTICS_4,
	DATE_RANGE_OFFSET as DATE_RANGE_OFFSET_ANALYTICS,
} from '../../../../analytics-4/datastore/constants';
import Chart from './Chart';

function SearchFunnelWidgetGA4( { Widget, WidgetReportError } ) {
	const [ selectedStats, setSelectedStats ] = useState( 0 );

	const viewOnly = useViewOnly();

	const isAnalytics4Available = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleAvailable( 'analytics-4' )
	);

	const canViewSharedAnalytics4 = useSelect( ( select ) => {
		if ( ! isAnalytics4Available ) {
			return false;
		}

		if ( ! viewOnly ) {
			return true;
		}

		return select( CORE_USER ).canViewSharedModule( 'analytics-4' );
	} );

	const isGA4Connected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'analytics-4' )
	);
	const dateRangeLength = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeNumberOfDays()
	);
	const url = useSelect( ( select ) =>
		select( CORE_SITE ).getCurrentEntityURL()
	);

	const { endDate, compareStartDate } = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			compare: true,
			offsetDays: DATE_RANGE_OFFSET,
		} )
	);
	const ga4Dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			compare: true,
			offsetDays: DATE_RANGE_OFFSET_ANALYTICS,
		} )
	);

	const showRecoverableAnalytics = useSelect( ( select ) => {
		if ( ! viewOnly ) {
			return false;
		}

		const recoverableModules =
			select( CORE_MODULES ).getRecoverableModules();

		if ( recoverableModules === undefined ) {
			return undefined;
		}

		return Object.keys( recoverableModules ).includes( 'analytics-4' );
	} );

	const ga4ConversionsData = useInViewSelect(
		( select ) => {
			return isGA4Connected &&
				canViewSharedAnalytics4 &&
				! showRecoverableAnalytics
				? select( MODULES_ANALYTICS_4 ).getConversionEvents()
				: [];
		},
		[ isGA4Connected, canViewSharedAnalytics4, showRecoverableAnalytics ]
	);

	const searchConsoleReportArgs = {
		startDate: compareStartDate,
		endDate,
		dimensions: 'date',
	};

	const ga4OverviewArgs = {
		...ga4Dates,
		metrics: [
			{
				name: 'conversions',
			},
			{
				name: 'engagementRate',
			},
		],
		dimensionFilters: {
			sessionDefaultChannelGrouping: [ 'Organic Search' ],
		},
	};

	const ga4StatsArgs = {
		...ga4Dates,
		...ga4OverviewArgs,
		dimensions: [
			{
				name: 'date',
			},
		],
		orderby: [
			{
				dimension: {
					dimensionName: 'date',
				},
			},
		],
	};
	const ga4VisitorsOverviewAndStatsArgs = {
		...ga4Dates,
		metrics: [
			{
				name: 'totalUsers',
			},
		],
		dimensions: [
			{
				name: 'date',
			},
		],
		dimensionFilters: {
			sessionDefaultChannelGrouping: [ 'Organic Search' ],
		},
		orderby: [
			{
				dimension: {
					dimensionName: 'date',
				},
			},
		],
	};

	if ( isURL( url ) ) {
		searchConsoleReportArgs.url = url;
		ga4OverviewArgs.url = url;
		ga4StatsArgs.url = url;
		ga4VisitorsOverviewAndStatsArgs.url = url;
	}

	const searchConsoleData = useInViewSelect(
		( select ) =>
			select( MODULES_SEARCH_CONSOLE ).getReport(
				searchConsoleReportArgs
			),
		[ searchConsoleReportArgs ]
	);
	const searchConsoleError = useSelect( ( select ) =>
		select( MODULES_SEARCH_CONSOLE ).getErrorForSelector( 'getReport', [
			searchConsoleReportArgs,
		] )
	);

	const searchConsoleLoading = useSelect(
		( select ) =>
			! select( MODULES_SEARCH_CONSOLE ).hasFinishedResolution(
				'getReport',
				[ searchConsoleReportArgs ]
			)
	);

	const ga4OverviewData = useInViewSelect(
		( select ) => {
			if (
				! isGA4Connected ||
				! canViewSharedAnalytics4 ||
				showRecoverableAnalytics
			) {
				return null;
			}

			return select( MODULES_ANALYTICS_4 ).getReport( ga4OverviewArgs );
		},
		[
			isGA4Connected,
			canViewSharedAnalytics4,
			showRecoverableAnalytics,
			ga4OverviewArgs,
		]
	);
	const ga4StatsData = useInViewSelect(
		( select ) => {
			if (
				! isGA4Connected ||
				! canViewSharedAnalytics4 ||
				showRecoverableAnalytics
			) {
				return null;
			}

			return select( MODULES_ANALYTICS_4 ).getReport( ga4StatsArgs );
		},
		[
			isGA4Connected,
			canViewSharedAnalytics4,
			showRecoverableAnalytics,
			ga4StatsArgs,
		]
	);
	const ga4VisitorsOverviewAndStatsData = useInViewSelect(
		( select ) => {
			if (
				! isGA4Connected ||
				! canViewSharedAnalytics4 ||
				showRecoverableAnalytics
			) {
				return null;
			}

			return select( MODULES_ANALYTICS_4 ).getReport(
				ga4VisitorsOverviewAndStatsArgs
			);
		},
		[
			isGA4Connected,
			canViewSharedAnalytics4,
			showRecoverableAnalytics,
			ga4VisitorsOverviewAndStatsArgs,
		]
	);

	const ga4Loading = useSelect( ( select ) => {
		if (
			! isGA4Connected ||
			! canViewSharedAnalytics4 ||
			showRecoverableAnalytics
		) {
			return false;
		}

		const { hasFinishedResolution } = select( MODULES_ANALYTICS_4 );

		return ! (
			hasFinishedResolution( 'getReport', [ ga4OverviewArgs ] ) &&
			hasFinishedResolution( 'getReport', [ ga4StatsArgs ] ) &&
			hasFinishedResolution( 'getReport', [
				ga4VisitorsOverviewAndStatsArgs,
			] ) &&
			hasFinishedResolution( 'getConversionEvents', [] )
		);
	} );

	const ga4Error = useSelect( ( select ) => {
		if ( ! isGA4Connected || showRecoverableAnalytics ) {
			return null;
		}

		const { getErrorForSelector } = select( MODULES_ANALYTICS_4 );

		return (
			getErrorForSelector( 'getReport', [ ga4OverviewArgs ] ) ||
			getErrorForSelector( 'getReport', [ ga4StatsArgs ] ) ||
			getErrorForSelector( 'getReport', [
				ga4VisitorsOverviewAndStatsArgs,
			] ) ||
			getErrorForSelector( 'getConversionEvents', [] )
		);
	} );

	const isGA4GatheringData = useInViewSelect(
		( select ) =>
			isGA4Connected &&
			canViewSharedAnalytics4 &&
			! showRecoverableAnalytics
				? select( MODULES_ANALYTICS_4 ).isGatheringData()
				: false,
		[ isGA4Connected, canViewSharedAnalytics4, showRecoverableAnalytics ]
	);
	const isSearchConsoleGatheringData = useInViewSelect( ( select ) =>
		select( MODULES_SEARCH_CONSOLE ).isGatheringData()
	);

	function WidgetFooter() {
		return (
			<Footer
				metrics={ SearchFunnelWidgetGA4.metrics }
				selectedStats={ selectedStats }
			/>
		);
	}

	if ( searchConsoleError ) {
		return (
			<Widget Header={ Header } Footer={ WidgetFooter }>
				<WidgetReportError
					moduleSlug="search-console"
					error={ searchConsoleError }
				/>
			</Widget>
		);
	}

	if (
		searchConsoleLoading ||
		ga4Loading ||
		isGA4GatheringData === undefined ||
		isSearchConsoleGatheringData === undefined
	) {
		return (
			<Widget Header={ Header } Footer={ WidgetFooter } noPadding>
				<PreviewBlock width="100%" height="190px" padding />
				<PreviewBlock width="100%" height="270px" padding />
			</Widget>
		);
	}

	return (
		<Widget noPadding Header={ Header } Footer={ WidgetFooter }>
			<Overview
				ga4Data={ ga4OverviewData }
				ga4ConversionsData={ ga4ConversionsData }
				ga4VisitorsData={ ga4VisitorsOverviewAndStatsData }
				searchConsoleData={ searchConsoleData }
				handleStatsSelection={ setSelectedStats }
				selectedStats={ selectedStats }
				dateRangeLength={ dateRangeLength }
				error={ ga4Error }
				WidgetReportError={ WidgetReportError }
				showRecoverableAnalytics={ showRecoverableAnalytics }
			/>

			<Chart
				canViewSharedAnalytics4={ canViewSharedAnalytics4 }
				dateRangeLength={ dateRangeLength }
				ga4StatsData={ ga4StatsData }
				ga4VisitorsOverviewAndStatsData={
					ga4VisitorsOverviewAndStatsData
				}
				isGA4Connected={ isGA4Connected }
				isGA4GatheringData={ isGA4GatheringData }
				isSearchConsoleGatheringData={ isSearchConsoleGatheringData }
				metrics={ SearchFunnelWidgetGA4.metrics }
				searchConsoleData={ searchConsoleData }
				selectedStats={ selectedStats }
				showRecoverableAnalytics={ showRecoverableAnalytics }
			/>
		</Widget>
	);
}

SearchFunnelWidgetGA4.metrics = [
	{
		id: 'impressions',
		color: '#6380b8',
		label: __( 'Impressions', 'google-site-kit' ),
		metric: 'impressions',
		service: 'search-console',
	},
	{
		id: 'clicks',
		color: '#4bbbbb',
		label: __( 'Clicks', 'google-site-kit' ),
		metric: 'clicks',
		service: 'search-console',
	},
	{
		id: 'users',
		color: '#3c7251',
		label: __( 'Users', 'google-site-kit' ),
		service: 'analytics-4',
	},
	{
		id: 'conversions',
		color: '#8e68cb',
		label: __( 'Conversions', 'google-site-kit' ),
		service: 'analytics-4',
	},
	{
		id: 'engagement-rate',
		color: '#8e68cb',
		label: __( 'Engagement Rate', 'google-site-kit' ),
		service: 'analytics-4',
	},
];

SearchFunnelWidgetGA4.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	WidgetReportZero: PropTypes.elementType.isRequired,
	WidgetReportError: PropTypes.elementType.isRequired,
};

export default SearchFunnelWidgetGA4;
