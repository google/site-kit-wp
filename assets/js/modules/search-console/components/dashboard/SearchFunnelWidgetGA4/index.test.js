/**
 * SearchFunnelWidgetGA4 component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
	render,
	createTestRegistry,
	provideModules,
	provideUserCapabilities,
	provideUserInfo,
	provideUserAuthentication,
	provideSiteInfo,
	muteFetch,
} from '../../../../../../../tests/js/test-utils';
import coreModulesFixture from '../../../../../googlesitekit/modules/datastore/__fixtures__';
import { CORE_MODULES } from '../../../../../googlesitekit/modules/datastore/constants';
import { CORE_USER } from '../../../../../googlesitekit/datastore/user/constants';
import { MODULES_SEARCH_CONSOLE } from '../../../datastore/constants';
import { getWidgetComponentProps } from '../../../../../googlesitekit/widgets/util';
import SearchFunnelWidgetGA4 from '.';
import {
	getViewportWidth,
	setViewportWidth,
} from '../../../../../../../tests/js/viewport-width-utils';

describe( 'SearchFunnelWidgetGA4', () => {
	let registry;
	let originalViewport;

	const widgetComponentProps = getWidgetComponentProps( 'searchFunnel' );

	beforeEach( () => {
		registry = createTestRegistry();

		provideModules( registry );
		provideUserInfo( registry );
		provideUserAuthentication( registry );
		provideUserCapabilities( registry );
		provideSiteInfo( registry );
		registry.dispatch( CORE_USER ).receiveConnectURL( 'test-url' );
		registry.dispatch( MODULES_SEARCH_CONSOLE ).receiveGetSettings( {
			propertyID: 'http://example.com/',
		} );
		registry.dispatch( CORE_USER ).setReferenceDate( '2021-01-28' );

		fetchMock.getOnce(
			new RegExp(
				'^/google-site-kit/v1/modules/search-console/data/searchanalytics'
			),
			{
				body: [
					{
						clicks: 123,
						ctr: 4.56,
						impressions: 7890,
						keys: [ '2022-06-21' ],
						position: 12.345,
					},
				],
			}
		);

		muteFetch(
			new RegExp(
				'^/google-site-kit/v1/modules/search-console/data/data-available'
			)
		);

		originalViewport = getViewportWidth();
		setViewportWidth( 450 );
	} );

	afterEach( () => {
		setViewportWidth( originalViewport );
	} );

	it( 'should render the Search Funnel Widget, including the Activate Analytics CTA', async () => {
		const { container, getByText, waitForRegistry } = render(
			<SearchFunnelWidgetGA4 { ...widgetComponentProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( /Set up Google Analytics/ ) ).toBeInTheDocument();
	} );

	it( 'should not render the Activate Analytics CTA when the Analytics module is not available', async () => {
		registry
			.dispatch( CORE_MODULES )
			.receiveGetModules(
				coreModulesFixture.filter(
					( { slug } ) => slug !== 'analytics-4'
				)
			);

		const { container, queryByText, waitForRegistry } = render(
			<SearchFunnelWidgetGA4 { ...widgetComponentProps } />,
			{ registry }
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect(
			queryByText( /Set up Google Analytics/ )
		).not.toBeInTheDocument();
	} );
} );
