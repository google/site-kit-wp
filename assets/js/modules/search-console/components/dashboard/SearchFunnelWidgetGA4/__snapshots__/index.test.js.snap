// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFunnelWidgetGA4 should not render the Activate Analytics CTA when the Analytics module is not available 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--searchFunnel googlesitekit-widget--no-padding googlesitekit-widget--with-header"
  >
    <div
      class="googlesitekit-widget__header"
    >
      <h3
        class="googlesitekit-subheading-1 googlesitekit-widget__header-title"
      >
        Search traffic over the last 28 days
      </h3>
    </div>
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone"
          >
            <div
              class="mdc-layout-grid__inner"
            >
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Impressions"
                  aria-pressed="true"
                  class="googlesitekit-data-block googlesitekit-data-block--impressions googlesitekit-data-block--button-1 googlesitekit-data-block--button googlesitekit-data-block--selected"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Impressions
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      7.9K
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Clicks"
                  aria-pressed="false"
                  class="googlesitekit-data-block googlesitekit-data-block--clicks googlesitekit-data-block--button-2 googlesitekit-data-block--button"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Clicks
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      123
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mdc-layout-grid googlesitekit-search-console-site-stats"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-chart googlesitekit-chart--LineChart"
              id="googlesitekit-chart-1"
              tabindex="-1"
            >
              <div
                class="googlesitekit-chart-loading"
              >
                <div
                  class="googlesitekit-preview-block googlesitekit-chart-loading__wrapper"
                  style="width: 100%; height: 270px;"
                >
                  <div
                    class="googlesitekit-preview-block__wrapper"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-widget__footer"
    >
      <div
        class="googlesitekit-source-link"
      >
        Source: 
        <a
          aria-label="Search Console (opens in a new tab)"
          class="googlesitekit-cta-link"
          href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fsearch.google.com%2Fsearch-console%2Fperformance%2Fsearch-analytics%3Fresource_id%3Dhttp%253A%252F%252Fexample.com%252F%26metrics%3Dimpressions%26start_date%3D20201231%26end_date%3D20210127&Email=wapuu.wordpress%40gmail.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Search Console
          </span>
          <span
            class="googlesitekit-icon-wrapper"
            style="margin-left: 5px;"
          >
            <svg />
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchFunnelWidgetGA4 should not render the Activate Analytics CTA when the Analytics module is not available 2`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--searchFunnel googlesitekit-widget--no-padding googlesitekit-widget--with-header"
  >
    <div
      class="googlesitekit-widget__header"
    >
      <h3
        class="googlesitekit-subheading-1 googlesitekit-widget__header-title"
      >
        Search traffic over the last 28 days
      </h3>
    </div>
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone"
          >
            <div
              class="mdc-layout-grid__inner"
            >
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Impressions"
                  aria-pressed="true"
                  class="googlesitekit-data-block googlesitekit-data-block--impressions googlesitekit-data-block--button-1 googlesitekit-data-block--button googlesitekit-data-block--selected"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Impressions
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      7.9K
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Clicks"
                  aria-pressed="false"
                  class="googlesitekit-data-block googlesitekit-data-block--clicks googlesitekit-data-block--button-2 googlesitekit-data-block--button"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Clicks
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      123
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mdc-layout-grid googlesitekit-search-console-site-stats"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-chart googlesitekit-chart--LineChart"
              id="googlesitekit-chart-1"
              tabindex="-1"
            >
              <div
                class="googlesitekit-chart-loading"
              >
                <div
                  class="googlesitekit-preview-block googlesitekit-chart-loading__wrapper"
                  style="width: 100%; height: 270px;"
                >
                  <div
                    class="googlesitekit-preview-block__wrapper"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-widget__footer"
    >
      <div
        class="googlesitekit-source-link"
      >
        Source: 
        <a
          aria-label="Search Console (opens in a new tab)"
          class="googlesitekit-cta-link"
          href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fsearch.google.com%2Fsearch-console%2Fperformance%2Fsearch-analytics%3Fresource_id%3Dhttp%253A%252F%252Fexample.com%252F%26metrics%3Dimpressions%26start_date%3D20201231%26end_date%3D20210127&Email=wapuu.wordpress%40gmail.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Search Console
          </span>
          <span
            class="googlesitekit-icon-wrapper"
            style="margin-left: 5px;"
          >
            <svg />
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchFunnelWidgetGA4 should render the Search Funnel Widget, including the Activate Analytics CTA 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--searchFunnel googlesitekit-widget--no-padding googlesitekit-widget--with-header"
  >
    <div
      class="googlesitekit-widget__header"
    >
      <h3
        class="googlesitekit-subheading-1 googlesitekit-widget__header-title"
      >
        Search traffic over the last 28 days
      </h3>
    </div>
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone"
          >
            <div
              class="mdc-layout-grid__inner"
            >
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Impressions"
                  aria-pressed="true"
                  class="googlesitekit-data-block googlesitekit-data-block--impressions googlesitekit-data-block--button-1 googlesitekit-data-block--button googlesitekit-data-block--selected"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Impressions
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      7.9K
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-2-phone"
              >
                <div
                  aria-label="Total Clicks"
                  aria-pressed="false"
                  class="googlesitekit-data-block googlesitekit-data-block--clicks googlesitekit-data-block--button-2 googlesitekit-data-block--button"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="googlesitekit-data-block__title-datapoint-wrapper"
                  >
                    <h3
                      class=" googlesitekit-subheading-1 googlesitekit-data-block__title "
                    >
                      <span
                        class="googlesitekit-data-block__title-inner"
                      >
                        Total Clicks
                      </span>
                    </h3>
                    <div
                      class="googlesitekit-data-block__datapoint"
                    >
                      123
                    </div>
                  </div>
                  <div
                    class="googlesitekit-data-block__change-source-wrapper"
                  >
                    <div
                      class="googlesitekit-data-block__change googlesitekit-data-block__change--no-change"
                    >
                      <span
                        class="googlesitekit-data-block__value"
                      >
                        0%
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-6-desktop mdc-layout-grid__cell--span-4-tablet mdc-layout-grid__cell--span-4-phone"
          />
        </div>
      </div>
      <div
        class="mdc-layout-grid googlesitekit-search-console-site-stats"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
          >
            <div
              class="googlesitekit-chart googlesitekit-chart--LineChart"
              id="googlesitekit-chart-0"
              tabindex="-1"
            >
              <div
                class="googlesitekit-chart-loading"
              >
                <div
                  class="googlesitekit-preview-block googlesitekit-chart-loading__wrapper"
                  style="width: 100%; height: 270px;"
                >
                  <div
                    class="googlesitekit-preview-block__wrapper"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mdc-layout-grid"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell"
          >
            <div
              class="googlesitekit-analytics-cta"
            >
              <div
                class="googlesitekit-analytics-cta__preview-graphs"
              >
                <div
                  class="googlesitekit-analytics-cta__preview-graph"
                >
                  <h3
                    class="googlesitekit-analytics-cta__preview-graph--title"
                  >
                    Unique visitors from Search
                  </h3>
                  <div>
                    <svg />
                  </div>
                  <div
                    class="googlesitekit-analytics-cta__preview-graph--icons"
                  >
                    <svg />
                    <span
                      class="googlesitekit-analytics-cta__preview-graph--bar"
                    />
                  </div>
                </div>
                <div
                  class="googlesitekit-analytics-cta__preview-graph"
                >
                  <h3
                    class="googlesitekit-analytics-cta__preview-graph--title"
                  >
                    Conversions completed
                  </h3>
                  <div>
                    <svg />
                  </div>
                  <div
                    class="googlesitekit-analytics-cta__preview-graph--icons"
                  >
                    <svg />
                    <span
                      class="googlesitekit-analytics-cta__preview-graph--bar"
                    />
                  </div>
                </div>
              </div>
              <div
                class="googlesitekit-analytics-cta__details"
              >
                <p
                  class="googlesitekit-analytics-cta--description"
                >
                  See how many people visit your site from Search and track how you’re achieving your goals
                </p>
                <button
                  class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
                  target="_self"
                >
                  <span
                    class="mdc-button__label"
                  >
                    Set up Google Analytics
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-widget__footer"
    >
      <div
        class="googlesitekit-source-link"
      >
        Source: 
        <a
          aria-label="Search Console (opens in a new tab)"
          class="googlesitekit-cta-link"
          href="https://accounts.google.com/accountchooser?continue=https%3A%2F%2Fsearch.google.com%2Fsearch-console%2Fperformance%2Fsearch-analytics%3Fresource_id%3Dhttp%253A%252F%252Fexample.com%252F%26metrics%3Dimpressions%26start_date%3D20201231%26end_date%3D20210127&Email=wapuu.wordpress%40gmail.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Search Console
          </span>
          <span
            class="googlesitekit-icon-wrapper"
            style="margin-left: 5px;"
          >
            <svg />
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;
